src/pool/fifo.c

```diff
@@ -169,6 +223,21 @@ static void pool_push_shared(ABT_pool pool, ABT_unit unit)
     ABTI_spinlock_release(&p_data->mutex);
 }
 
+static void pool_push_shared_blocking(ABT_pool pool, ABT_unit unit)
+{
+    ABTI_pool *p_pool = ABTI_pool_get_ptr(pool);
+    void *data = ABTI_pool_get_data(p_pool);
+    data_t *p_data = pool_get_data_ptr(data);
+
+    /* hold additional pthread mutex and signal anyone who might be blocking
+     * on pop()
+     */
+    pthread_mutex_lock(&p_data->blocking_mutex);
```

Comment:
I don't think acquiring the lock here is necessary. The pool push operation should be thread safe and `cond_signal` does not need a lock

Reply:
I agree cond_signal() itself doesn't need a lock, but the lock in this specific example is to prevent a race between the waiter checking for the pool being empty and this function adding something to the pool.  Without the mutex the you could theoretically get this order of operations, with thread a calling pop() and thread b calling push():

thread a: lock 
thread a: check if pool is empty
thread b: insert into pool queue
thread b: cond_signal
thread a: cond_wait

Thread a will block in this case even though there is work in the pool when it hits the cond_wait() call.

Reply:
the timeout != infinite avoids deadlocks in this PR, so I think it's still correct. Performance wise, my suggestion allows push() to be on a fast path and pop() on empty (including false positives) on a slow path. But I agree that if we change the semantic of the pool to include infinite wait on pop() (which I think is a likely API extension in the future), then keeping the lock for correctness is a must